File: app/src/components/main_page.rs
use yew::prelude::*;

use super::{Button, PasswordList, Knowledgebase};

#[derive(PartialEq)]
pub enum Views {
    PasswordList,
    Knowledge,
}

pub enum Messages {
    ChangeView(Views),
}

pub struct MainPage {
    link: ComponentLink<Self>,
    view: Views,
}

impl Component for MainPage {
    type Message = Messages;
    type Properties = ();

    fn create(_: Self::Properties, link: ComponentLink<Self>) -> Self {
        Self {
            link,
            view: Views::PasswordList,
        }
    }

    fn update(&mut self, msg: Self::Message) -> ShouldRender {
        match msg {
            Messages::ChangeView(view) => {
                self.view = view;
                true
            },
        }
    }

    fn change(&mut self, _: Self::Properties) -> ShouldRender {
        false
    }

    fn view(&self) -> Html {
        let password_active = self.view == Views::PasswordList;
        let knowledge_active = self.view == Views::Knowledge;

        html! {
            <div class="main-layout main-background animation-fade">
                <aside class="main-menu">
                    <img class="main-menu-logo" src="icons/brain.svg" alt="" />

                    <Button active=password_active clicked=self.link.callback(|_| Messages::ChangeView(Views::PasswordList))>
                        <img src="icons/password.svg" alt="Passwords" />
                    </Button>

                    <Button active=knowledge_active clicked=self.link.callback(|_| Messages::ChangeView(Views::Knowledge))>
                        <img src="icons/knowledge.svg" alt="Knowledge" />
                    </Button>
                </aside>
                <section class="main-content">
                {
                    match self.view {
                        Views::PasswordList => html! {
                            <PasswordList />
                        },
                        Views::Knowledge => html! {
                            <Knowledgebase />
                        },
                    }
                }
                </section>
            </div>
        }
    }
}

File: src/pages/main/mainPage.js
import { Component } from 'react';

import { Button } from '../../components/button';

import { Logo } from '../../icons/logo';
import { Passwords } from '../../icons/passwords';
import { Knowledge } from '../../icons/knowledge';

import { HomePage } from '../../pages/home/homePage';
import { KnowledgePage } from '../../pages/knowledge/knowledgePage';
import { LoginPage } from '../../pages/login/loginPage';
import { PasswordsPage } from '../../pages/passwords/passwordsPage';
import { SyncPage } from '../../pages/sync/syncPage';

import { loadKey } from '../../services/crypto';

import './main.css';

export class MainPage extends Component {
	constructor(props) {
		super(props);

		this.state = {
			view: "home",
		};
	}

	render() {
		const [loggedIn, ] = loadKey();
		if (!loggedIn) {
			return (
				<LoginPage changeView={view => this.setState({ view: view })} />
			);
		}

		return (
            <div className="main-layout main-background">
                <aside className="main-menu">
                    <div className="menu-items">
                        <Logo class="main-menu-logo" clicked={() => this.setState({ view: "home"})} />

                        <Button active={this.state.view === "passwords"} clicked={() => this.setState({ view: "passwords"})}>
                            <Passwords class="main-button-icon" />
                        </Button>
                        <Button active={this.state.view === "knowledge"} clicked={() => this.setState({ view: "knowledge"})}>
                            <Knowledge class="main-button-icon" />
                        </Button>
                    </div>
                </aside>
                <section className="main-content">
					{ this.state.view === "passwords" &&
						<PasswordsPage changeView={view => this.setState({ view: view })} />
					}
					{ this.state.view === "knowledge" &&
						<KnowledgePage changeView={view => this.setState({ view: view })} />
					}
					{ this.state.view === "sync" &&
						<SyncPage changeView={view => this.setState({ view: view })} />
					}
					{ this.state.view === "home" &&
						<HomePage changeView={view => this.setState({ view: view })} />
					}
                </section>
            </div>
		);
	}
}